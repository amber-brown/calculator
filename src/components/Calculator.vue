<template>
    <div class="calculator">
      <div class="calculator__display">
        <p class="calculator__display-text">{{ input }}</p>
      </div>
      <div class="calculator__buttons-wrapper">
        <div v-for="(buttonRow, index) in buttons" :key="index" class="calculator__button-row">
          <button
            v-for="button in buttonRow"
            @click="onButtonPress(button)"
            :key="button"
            class="calculator__button"
          >
              {{button}}
          </button>
        </div>
      </div>
    </div>
</template>

<script>
import * as calculator from '@/utils/calculator';

export default {
  name: 'Calculator',
  data() {
    return {
      buttons: [
        ['7', '8', '9', '+'],
        ['4', '5', '6', '-'],
        ['1', '2', '3', '*'],
        ['c', '0', '=', '/'],
      ],
      input: '',
    };
  },
  computed: {
    tokens() {
      return calculator.tokenizer(this.input);
    },
    lastToken() {
      return this.tokens[this.tokens.length - 1];
    },
  },
  methods: {
    onButtonPress(button) {
      if (button === 'c') {
        this.input = '';
        return;
      }

      if (button === '=') {
        if (this.lastToken?.type === 'Operator') {
          this.input = this.input.slice(0, -1);
          // TODO : test
        }
        this.onEquate();
        return;
      }

      if (!this.input.length && calculator.isOperator(button)) {
        return;
      }

      if (calculator.isDigit(button) && this.lastToken?.value === '0') {
        this.input = `${this.input.slice(0, -1)}${button}`;
        return;
      }

      if (calculator.isOperator(button) && this.lastToken?.type === 'Operator') {
        this.input = `${this.input.slice(0, -1)}${button}`;
        return;
      }
      this.input = `${this.input}${button}`;
    },
    onEquate() {
      console.log(this.input);
    },
  },
};
</script>

<style>
.calculator {
  padding: 8px;
}

.calculator__display {
  border: 1px solid grey;
  border-radius: 6px;
  margin: 6px;
  padding: 6px 12px;
  height: 50px;
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.calculator__display-text {
  display: inline;
  font-size: 18px;
  margin: 0;
}

.calculator__button-row {
  display: flex;
}

.calculator__button {
  margin: 6px;
  padding: 6px;
  border-radius: 6px;
  border: none;
  outline: none;
  width: 100%;
  height: 50px;
}
</style>
