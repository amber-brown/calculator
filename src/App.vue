<template>
  <div id="app">
    <div class="calculator-container">
    <Calculator />
    </div>
  </div>
</template>

<script>
import Calculator from '@/components/Calculator.vue';

export default {
  name: 'App',
  components: {
    Calculator,
  },
};
</script>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
  margin-top: 60px;
}

.calculator-container {
  max-width: 700px;
  margin: 0 auto;
}
</style>
