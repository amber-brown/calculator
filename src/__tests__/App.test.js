import Calculator from '@/components/Calculator.vue';
import { shallowMount } from '@vue/test-utils';
import App from '../App.vue';

let wrapper;

describe('App', () => {
  beforeEach(() => {
    wrapper = shallowMount(App);
  });

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('displays the calculator', () => {
    const calculator = wrapper.findComponent(Calculator);
    expect(calculator.exists()).toBeTruthy();
  });
});
