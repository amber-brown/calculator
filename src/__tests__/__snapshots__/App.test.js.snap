// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  id="app"
>
  <div
    class="calculator-container"
  >
    <calculator-stub />
  </div>
</div>
`;
